runtime/src/bank.rs

```diff
@@ -5011,39 +5011,55 @@ impl Bank {
                     .collect()
             };
 
-        let ExtractedPrograms {
-            loaded: mut loaded_programs_for_txs,
-            missing,
-        } = {
-            // Lock the global cache to figure out which programs need to be loaded
-            let loaded_programs_cache = self.loaded_programs_cache.read().unwrap();
-            Mutex::into_inner(
-                Arc::into_inner(
-                    loaded_programs_cache.extract(self.slot, programs_and_slots.into_iter()),
-                )
-                .unwrap(),
-            )
-            .unwrap()
-        };
-
-        // Load missing programs while global cache is unlocked
-        let missing_programs: Vec<(Pubkey, Arc<LoadedProgram>)> = missing
-            .iter()
-            .map(|(key, (count, reloading))| {
-                let program = self.load_program(key, *reloading, None);
-                program.tx_usage_counter.store(*count, Ordering::Relaxed);
-                (*key, program)
-            })
-            .collect();
+        let mut loaded_programs_for_txs = None;
+        let mut program_to_store = None;
+        loop {
+            let (program_to_load, task_cookie, task_waiter) = {
+                // Lock the global cache.
+                let mut loaded_programs_cache = self.loaded_programs_cache.write().unwrap();
+                // Initialize our local cache.
+                if loaded_programs_for_txs.is_none() {
+                    loaded_programs_for_txs = Some(LoadedProgramsForTxBatch::new(
```

Comment:
Can this be initialized outside the loop? It's only being set once. That'll also remove the need for an `Option`.

Reply:
The option is on purpose so that we don't need to lock the `LoadedPrograms` once outside the loop just to get the current environment.

Reply:
Yes, the lock requirement was understood. 

I think it comes down to saving on one `read` lock, vs checking if the variable is initialized on every iteration of the loop. What worried me the most was `unwrap()` at the end of the function for the return value. Current code is initializing the `Option` in the first iteration of the loop, and it works. Some future change could potentially change that bahavior.

Maybe I am overthinking it.